docs/_posts/2021-03-01-hudi-file-sizing.md

```diff
@@ -0,0 +1,96 @@
+
+The algorithm is generally applicable to any partitions and so our illustration is just going to focus on a single 
+partition only. Let’s say this is the layout of data files for a given partition.
+
+![Initial layout](/assets/images/blog/hudi-file-sizing/initial_layout.png)
+_Figure: Initial data file sizes for a given partition of interest_
+
+Let’s assume the configured values for max file size and small file size limit is 120Mb and 100Mb. File_1’s current 
+size is 40Mb, File_2’s size is 80Mb, File_3’s size is 90Mb, File_4’s size is 130Mb and File_5’s size is 105Mb.
+
+Let’s see what happens when a new write batch is ingested to hudi. This is done in multiple steps. First, updates are 
+assigned to respective files followed which inserts are assgined with the data files.
+
+Step1: Assigning updates to files. In this step, index is looked up to find the tagged location and records are 
+assigned to respective files. Note: Updates are not going to alter the file size as unique records per file will remain
+the same.
+
+Step2:  Determine small files for the partition of interest. The soft file limit config value will be leveraged here 
+to determine what constitutes as a small file. Given the config value is set to 100Mb, the small files are File_1(40Mb)
+and File_2(80Mb) and file_3’s (90Mb). No new inserts will go into any of the other existing files as they don’t qualify 
+for small files.
+
+Stpe3: Once small files are determined, incoming inserts are assigned to them so that they reach their max capacity of 
+120Mb. File_1 will be ingested with 80Mb worth of inserts, file_2 will be ingested with 40Mb worth of inserts and 
+File_3 will be ingested with 30Mb worth of inserts.
+
+![Bin packing small files](/assets/images/blog/hudi-file-sizing/bin_packing_existing_data_files.png)
+_Figure: Incoming records are bin packed to existing small files_
+
+Step4: Once all small files are bin packed to its max capacity and if there are pending inserts unassigned, new file 
+Ids are created and inserts are assigned to them. Number of Records per new data file is determined from insert split 
+size config. Assuming the insert split size is configured to 120k records, if there are 300k remaining records, 3 new 
+files will be created in which 2 of them (File_6 and File_7) will be filled with 120k records and the last one (File_8)
+will be filled with 60k records (assuming each record is 1000 bytes). In future ingestions, 3rd new file will be 
+considered as a small file to be packed with more data.
+
+![Assigning to new files](/assets/images/blog/hudi-file-sizing/adding_new_files.png)
+_Figure: Remaining records are assigned to new files_
+
+Hudi has a custom partitioner which will be leveraged by the execution engine and records will be distributed according
+to the assignment done as per the algorithm shown above.
+
+After this ingestion is complete, except File_8, every other file is nicely sized to its optimum size. This process is
```

Comment:
After this round of ingestion is complete, all files except File_8 are nicely sized to the optimum size. This process is
followed during every ingestion to ensure there are no small files in your Hudi data lake. This will help keep your read SLA's bounded.